src/mds/Server.cc

```diff
@@ -2511,19 +2511,32 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
   }
   
   if (is_full) {
-    if (req->get_op() == CEPH_MDS_OP_SETLAYOUT ||
-        req->get_op() == CEPH_MDS_OP_SETDIRLAYOUT ||
-        req->get_op() == CEPH_MDS_OP_SETLAYOUT ||
-        req->get_op() == CEPH_MDS_OP_RMXATTR ||
-        req->get_op() == CEPH_MDS_OP_SETXATTR ||
-        req->get_op() == CEPH_MDS_OP_CREATE ||
-        req->get_op() == CEPH_MDS_OP_SYMLINK ||
-        req->get_op() == CEPH_MDS_OP_MKSNAP ||
-	((req->get_op() == CEPH_MDS_OP_LINK ||
-	  req->get_op() == CEPH_MDS_OP_RENAME) &&
-	 (!mdr->has_more() || mdr->more()->witnessed.empty())) // haven't started peer request
+    CInode *cur = try_get_auth_inode(mdr, req->get_filepath().get_ino());
+    if (!cur) {
+      respond_to_request(mdr, -EINVAL);
+      return;
+    }
+    if ((req->get_op() == CEPH_MDS_OP_SETXATTR ||
+       req->get_op() == CEPH_MDS_OP_RENAME ||
+       req->get_op() == CEPH_MDS_OP_MKDIR ||
+       req->get_op() == CEPH_MDS_OP_RMDIR ||
+       req->get_op() == CEPH_MDS_OP_SYMLINK ||
+       req->get_op() == CEPH_MDS_OP_LINK ||
+       req->get_op() == CEPH_MDS_OP_UNLINK) &&
+       check_access(mdr, cur, MAY_FULL)) {
```

Comment:
This doesn't look right, shouldn't this `check_access` occur in the next `else if` statement? Client should always be able to free space. I think this causes:

https://tracker.ceph.com/issues/50870